### PR TITLE
further refactoring

### PR DESCRIPTION
**Pull request by rhaschke**
- display solution including scene / robot state
- receive num of solutions
- basic framework for editing a task in rviz
- split repo into 3 ROS packages

### CODE REVIEW

**Comment by v4hn** [comment]
Looks good, I added a minor work-around you missed while "cleaning up" my commits.
This branch segfaults rviz over here on qt4. Can you make sense of this backtrace?

```
#0  0x00007ffff5f5fa27 in raise () at /lib/libc.so.6
#1  0x00007ffff5f60c39 in abort () at /lib/libc.so.6
#2  0x00007ffff5f90a1f in __libc_message () at /lib/libc.so.6
#3  0x00007ffff5f95397 in malloc_printerr () at /lib/libc.so.6
#4  0x00007ffff5f960ac in _int_free () at /lib/libc.so.6
#5  0x00007fffb580782e in QModelIndex::parent() const () at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#6  0x00007fffb580d66a in moveit_rviz_plugin::TaskListModel::parent(QModelIndex const&) const ()
    at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#7  0x00007fffb580782e in QModelIndex::parent() const () at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#8  0x00007fffb5831448 in moveit_rviz_plugin::TaskDisplay::onTaskDataChanged(QModelIndex const&, QModelIndex const&) ()
    at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#9  0x00007fffb58641ed in moveit_rviz_plugin::TaskDisplay::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) ()
    at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#10 0x00007ffff2459dfb in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () at /usr/lib/qt4/libQtCore.so.4
#11 0x00007ffff248ad9a in QAbstractItemModel::dataChanged(QModelIndex const&, QModelIndex const&) () at /usr/lib/qt4/libQtCore.so.4
#12 0x00007fffb580f23e in moveit_rviz_plugin::TaskListModelPrivate::_q_sourceDataChanged(QModelIndex const&, QModelIndex const&) ()
    at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#13 0x00007fffb580f6bc in moveit_rviz_plugin::TaskListModel::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) ()
    at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#14 0x00007ffff2459dfb in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () at /usr/lib/qt4/libQtCore.so.4
#15 0x00007ffff248ad9a in QAbstractItemModel::dataChanged(QModelIndex const&, QModelIndex const&) () at /usr/lib/qt4/libQtCore.so.4
#16 0x00007fffb5807366 in moveit_rviz_plugin::RemoteTaskModel::processStageStatistics(std::vector<moveit_task_constructor_msgs::StageStatistics_<std::allocator<void> >, std::allocator<moveit_task_constructor_msgs::StageStatistics_<std::allocator<void> > > > const&) ()
    at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#17 0x00007fffb580df24 in moveit_rviz_plugin::TaskListModel::processTaskStatisticsMessage(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, moveit_task_constructor_msgs::TaskStatistics_<std::allocator<void> > const&) ()
    at devel/lib/libmotion_planning_tasks_rviz_plugin.so
#18 0x00007fffb58301fd in moveit_rviz_plugin::TaskDisplay::taskStatisticsCB(ros::MessageEvent<moveit_task_constructor_msgs::TaskStatistics_<std::allocator<void> > const> const&)::{lambda()#1}::operator()() const () at devel/lib/libmotion_planning_tasks_rviz_plugin.so
```

**Comment by v4hn** [comment]
Here's a Debug build backtrace.

Apparently execution hits https://github.com/ubi-agni/moveit_task_constructor/blob/720c08354eecd6b51e5b87085fe2a5a6b2a17257/visualization/motion_planning_tasks/src/task_list_model.cpp#L142 .

```
#0  0x00007ffff5f5fa27 in raise () at /lib/libc.so.6
#1  0x00007ffff5f60c39 in abort () at /lib/libc.so.6
#2  0x00007ffff23a8e86 in qt_message_output(QtMsgType, char const*) () at /usr/lib/qt4/libQtCore.so.4
#3  0x00007ffff23a8f94 in  () at /usr/lib/qt4/libQtCore.so.4
#4  0x00007ffff23a967a in qFatal(char const*, ...) () at /usr/lib/qt4/libQtCore.so.4
#5  0x00007fffb580ecf8 in moveit_rviz_plugin::TaskListModelPrivate::mapToSource(QModelIndex const&, moveit_rviz_plugin::TaskListModelPrivate::BaseModelData**)
const (this=0x21bd920, proxy_index=..., task=0x7fffffffc698)
    at src/moveit_task_constructor/visualization/motion_planning_tasks/src/task_list_model.cpp:142
#6  0x00007fffb580add1 in moveit_rviz_plugin::TaskListModel::parent(QModelIndex const&) const (this=0x12f6240, child=...)
    at src/moveit_task_constructor/visualization/motion_planning_tasks/src/task_list_model.cpp:242
#7  0x00007fffb5804f2e in QModelIndex::parent() const (this=0x7fffffffc940) at /usr/include/qt4/QtCore/qabstractitemmodel.h:393
#8  0x00007fffb582f0b4 in moveit_rviz_plugin::TaskDisplay::onTaskDataChanged(QModelIndex const&, QModelIndex const&) (this=0x1385350, topLeft=..., bottomRight=
...) at src/moveit_task_constructor/visualization/motion_planning_tasks/src/task_display.cpp:254
#9  0x00007fffb5861f0a in moveit_rviz_plugin::TaskDisplay::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_o=0x1385350, _c=QMetaObject::InvokeMe
taMethod, _id=4, _a=0x7fffffffc8d8)
    at build/moveit_task_constructor/visualization/motion_planning_tasks/src/motion_planning_tasks_rviz_plugin_autogen/EWIEGA46WW/moc_task_display.cpp:72
#10 0x00007ffff2459dfb in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () at /usr/lib/qt4/libQtCore.so.4
#11 0x00007ffff248ad9a in QAbstractItemModel::dataChanged(QModelIndex const&, QModelIndex const&) () at /usr/lib/qt4/libQtCore.so.4
#12 0x00007fffb580cb28 in moveit_rviz_plugin::TaskListModelPrivate::_q_sourceDataChanged(QModelIndex const&, QModelIndex const&) (this=0x21bd920, topLeft=...,
bottomRight=...) at src/moveit_task_constructor/visualization/motion_planning_tasks/src/task_list_model.cpp:568
#13 0x00007fffb580cfdb in moveit_rviz_plugin::TaskListModel::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) (_o=0x12f6240, _c=QMetaObject::Invoke
MetaMethod, _id=6, _a=0x7fffffffca88)
    at build/moveit_task_constructor/visualization/motion_planning_tasks/src/motion_planning_tasks_rviz_plugin_autogen/include/moc_task_list_model.cpp:129
#14 0x00007ffff2459dfb in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) () at /usr/lib/qt4/libQtCore.so.4
#15 0x00007ffff248ad9a in QAbstractItemModel::dataChanged(QModelIndex const&, QModelIndex const&) () at /usr/lib/qt4/libQtCore.so.4
#16 0x00007fffb58049ce in moveit_rviz_plugin::RemoteTaskModel::processStageStatistics(std::vector<moveit_task_constructor_msgs::StageStatistics_<std::allocator
<void> >, std::allocator<moveit_task_constructor_msgs::StageStatistics_<std::allocator<void> > > > const&) (this=0x24d0f70, msg=...)
    at src/moveit_task_constructor/visualization/motion_planning_tasks/src/remote_task_model.cpp:271
#17 0x00007fffb580b6d0 in moveit_rviz_plugin::TaskListModel::processTaskStatisticsMessage(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocat
or<char> > const&, moveit_task_constructor_msgs::TaskStatistics_<std::allocator<void> > const&) (this=0x12f6240, id=..., msg=...)
    at src/moveit_task_constructor/visualization/motion_planning_tasks/src/task_list_model.cpp:353
```

File: visualization/motion_planning_tasks/src/task_display.cpp [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+	if (isEnabled())
+		reset();
+	else
+		loadRobotModel();
+}
+
+void TaskDisplay::taskDescriptionCB(const ros::MessageEvent<const moveit_task_constructor_msgs::TaskDescription> &event)
+{
+	const moveit_task_constructor_msgs::TaskDescriptionConstPtr& msg = event.getMessage();
+	const std::string id = event.getPublisherName() + "/" + msg->id;
```
Comment (v4hn) [inline visualization/motion_planning_tasks/src/task_display.cpp]: This approach fails the moment you replay data, right?

Would it be better to extract the first part of your id from the topic the display subscribes to?

Do we really need this prefix in the id here?

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+notifications:
+  email:
+    recipients:
+      - rhaschke@techfak.uni-bielefeld.de
+      - me@v4hn.de
+env:
+  matrix:
+    - ROS_DISTRO=kinetic  ROS_REPO=ros
+    - ROS_DISTRO=kinetic  ROS_REPO=ros-shadow-fixed
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply .travis.yml]: I was trying to do so. However, as we require C++14, the Indigo container needs to be pimped with a decent gcc. If you want to take care of this, feel free!

File: core/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+catkin_package(
+	LIBRARIES
+		moveit_task_constructor_core
+		moveit_task_constructor_core_stages
+	INCLUDE_DIRS
+		include
+)
+
+set(CMAKE_CXX_STANDARD 14)
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply core/CMakeLists.txt]: No problem with this. However, I suggest to use a version check to switch between old/deprecated and new behaviour. I will add a corresponding patch.

Despite this marginal issue, I have lots of issues compiling against MoveIt Indigo from the ci container due to:
- `std::shared_ptr` vs. `boost::shared_ptr`
- `moveit/move_group_interface/move_group_interface.h` vs `moveit/move_group_interface/move_group.h`

How can you build this on Indigo at all?

File: msgs/msg/SubTrajectory.msg [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# unique id within task
+uint32 id
+
+# associated cost
+float32 cost
+
+# optional associated name of sub trajectory
+string name
```
[the remark this answers is not in the document]
Comment (rhaschke) [reply msgs/msg/SubTrajectory.msg]: For example we want to name different grasp solutions.

File: visualization/motion_planning_tasks/src/task_display.cpp (hunk 1 above)
Comment (rhaschke) [reply visualization/motion_planning_tasks/src/task_display.cpp]: The publisher name _is_ the full topic name, at least it was in my test.

**Comment by rhaschke** [comment]
@v4hn I reworked the item models, factoring out generic from application-specific stuff.
The unittests better cover some corner cases now and I hope, you don't see this rviz plugin crash anymore.
I also implemented solution navigation now. Please have a look.

**Comment by v4hn** [comment]
> However, as we require C++14, the Indigo container needs to be pimped with a decent gcc. If you want to take care of this, feel free!

> How can you build this on Indigo at all?

> https://github.com/TAMS-Group/moveit_task_constructor/pull/8/commits/40cfceb55788a23f4b03198147b57584130ed430

I don't build on indigo *on ubuntu 14.04* usually.
But I want to deploy on our PR2, of course, and that runs ubuntu 14.04 / indigo for the foreseeable future...
Also I believe there are other state-of-the-art (hoehoe) systems out there still running indigo.
It should be possible to build it there without *too* much trouble.

I will try to figure something out that works for me locally and is as little as possible a burden on the code base here.

File: msgs/msg/SubTrajectory.msg (hunk 4 above)
[the remark this answers is not in the document]
Comment (v4hn) [reply msgs/msg/SubTrajectory.msg]: Hm, this could be considered as a special-case of the `any` properties of InterfaceStates.
We did not discuss how they fit into the message structure yet. I believe it makes a lot of sense to publish them for debugging why stages miss parameters or received wrong parameters..

If you prefer to have an additional name field here, that's ok with me.

File: visualization/motion_planning_tasks/src/task_display.cpp (hunk 1 above)
Comment (v4hn) [reply visualization/motion_planning_tasks/src/task_display.cpp]: The inline documentation of `getPublisherName` says

> Returns the name of the node which published this message

To the best of my knowledge that's not the full topic.
If you get the topic name there and plan to use it, you should open an issue in ros_comm and sort out what *should* come out of this function.
I would prefer to avoid using `ros::MessageEvent` in the first place and get the topic name from the display property.

**Comment by v4hn** [comment]
> I will try to figure something out that works for me locally and is as little as possible a burden on the code base here.

So I tested some options and I agree that it is not reasonable to provide support for ROS indigo on Ubuntu 14.04 anymore. It's a lot of ridiculous overhead.
The compiler is severely outdated even w.r.t. C++11, not to talk about 14 (1y).
It is, however, not too hard to get a current gcc installed there. The [ubuntu toolchain ppa](https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test) provides working packages for gcc 6.1.
Additionally there is semi-official support for MoveIt's `kinetic-devel` branch for Ubuntu 14.04. (see e.g. [here](https://github.com/ros-planning/moveit/pull/682)).

With both these facts exploited, I can build and run the current `wip-refactor ` branch on 14.04.
The segfault I hit seems to be fixed too.
I believe the challenge to set this up reflects the fact that systems should seriously consider to upgrade to kinetic by now.
Even though I will probably put a tutorial/script online to do this soon.

Does that sound like a reasonable lower bound for the build requirements? I know things could be simplified quite a bit with making the hard lower bound 16.04/kinetic, but I'm not ready for that - and I believe neither are many other labs....

**Comment by rhaschke** [comment]
@v4hn I'm fine, if we can keep with C++14 (which you introduced originally). I also installed gcc 5 from the ppa you mentioned in my docker container to test building - which finally worked. Sticking with cmake 2.8 is OK as well. Supporting both Qt4 and Qt5 is painful but doable ;-)

Please don't (yet) autoformat with clang before you push to ros-planning. We should postpone that until things settled a little bit more (and I have merged all my loose ends).

**Comment by v4hn** [comment]
Thank you!

> Please don't (yet) autoformat with clang before you push to ros-planning.

Of course.
